python/example_code/devicefarm/device_testing/run_tests.py

```diff
@@ -0,0 +1,130 @@
+print(our_upload_arn, our_test_package_arn)
+# Now that we have those out of the way, we can start the test run...
+response = client.schedule_run(
+    projectArn = config["projectArn"],
+    appArn = our_upload_arn,
+    devicePoolArn = config["poolArn"],
+    name=unique,
+    test = {
+        "type":"APPIUM_PYTHON",
+        "testSpecArn": config["testSpecArn"],
+        "testPackageArn": our_test_package_arn
+        }
+    )
+run_arn = response['run']['arn']
+start_time = datetime.datetime.now()
+print(f"Run {unique} is scheduled as arn {run_arn} ")
+
+try:
+
+    while True:
+        response = client.get_run(arn=run_arn)
+        state = response['run']['status']
+        if state == 'COMPLETED' or state == 'ERRORED':
+            break
+        else:
+            print(f" Run {unique} in state {state}, total time "+str(datetime.datetime.now()-start_time))
+            time.sleep(10)
+except:
+    # If something goes wrong in this process, we stop the run and exit. 
+
+    client.stop_run(arn=run_arn)
+    exit(1)
+print(f"Tests finished in state {state} after "+str(datetime.datetime.now() - start_time))
+# now, we pull all the logs.
+jobs_response = client.list_jobs(arn=run_arn)
+# Save the output somewhere. We're using the unique value, but you could use something else
+save_path = os.path.join(os.getcwd(),'results', unique)
+os.mkdir(save_path)
+# Save the last run information
+for job in jobs_response['jobs'] :
+    # Make a directory for our information
+    job_name = job['name']
+    os.makedirs(os.path.join(save_path, job_name), exist_ok=True)
+    # Get each suite within the job
+    suites = client.list_suites(arn=job['arn'])['suites']
+    for suite in suites:
+        for test in client.list_tests(arn=suite['arn'])['tests']:
+            # Get the artifacts
+            for artifact_type in ['FILE','SCREENSHOT','LOG']:
+                artifacts = client.list_artifacts(
+                    type=artifact_type,
+                    arn = test['arn']
+                )['artifacts']
+                for artifact in artifacts:
+                    # We replace : because it has a special meaning in Windows & macos
+                    path_to = os.path.join(save_path, job_name, suite['name'], test['name'].replace(':','_') )
+                    os.makedirs(path_to, exist_ok=True)
+                    filename = artifact['type']+"_"+artifact['name']+"."+artifact['extension']
+                    artifact_save_path = os.path.join(path_to, filename)
+                    print("Downloading "+artifact_save_path)
+                    with open(artifact_save_path, 'wb') as fn, requests.get(artifact['url'],allow_redirects=True) as request:
+                        fn.write(request.content)
+                    #/for artifact in artifacts
+                #/for artifact type in []
+            #/ for test in ()[]
+        #/ for suite in suites
+    #/ for job in _[]
+# done
```

Comment:
I'd remove these trailing comments to reduce clutter--most IDEs will calculate your context for you.